### PR TITLE
xsdk: version: 1.0.0

### PR DESCRIPTION
**Pull request by balay**
~Note: This PR is currently for testing in spack CI (and is marked as `draft`)~

~Edit: `draft` prevents CI :(  so the draft status is removed in favor of `don't-merge-yet` label~

~Note: this PR is testing multiple changes to individual packages - that are to be merged to spack via separate PRs. And this branch will be rebased - and should eventually have one commit for `xsdk@1.0.0` - before merge to spack/develop~

Edit: This PR is ready for review/merge

Edit: additional xsdk@1.0.0 updates at https://github.com/spack/spack/pull/41180

### CODE REVIEW

File: var/spack/repos/builtin/packages/xsdk/package.py
Comment (mkstoyanov) [inline var/spack/repos/builtin/packages/xsdk/package.py]: We can remove `+xsdkflags`, I only keep it for backwards compatibility, otherwise it has no effect.

Comment (balay) [reply var/spack/repos/builtin/packages/xsdk/package.py]: thanks! removed now

**Review by tldahlgren — DISMISSED** [review]
Confirmed the affected versions, branches, and commits.  Will defer to maintainers for merge decision.

**Comment by balay** [comment]
https://gitlab.spack.io/spack/spack/-/pipelines/538785

CI has 2 issues 

- petsc: https://gitlab.spack.io/spack/spack/-/pipelines/538785
```
==> Installing petsc-3.20.1-z4wmr4b6xgxs5nkoqx62nh4jeqrs2pza [1/1]
==> Error: timeout: The read operation timed out
```
I don't understand this error. cc: @eugeneswalker 

- dealii:  https://gitlab.spack.io/spack/spack/-/jobs/8961187

Presumably will be fixed by: https://github.com/spack/spack/pull/40747

**Comment by eugeneswalker** [comment]
> CI has 2 issues
> 
>     * petsc: https://gitlab.spack.io/spack/spack/-/pipelines/538785
> 
> 
> ```
> ==> Installing petsc-3.20.1-z4wmr4b6xgxs5nkoqx62nh4jeqrs2pza [1/1]
> ==> Error: timeout: The read operation timed out
> ```

We are seeing this error exact error, also with `petsc@3.20.1` jobs, in a number of other pipelines I am watching right now -- I don't know what is going on and why we are seeing it specifically with petsc. @alalzo I know you noticed this error in another pipeline today, have you seen it with any other packages? Do you know what is going on?

**Comment by eugeneswalker** [comment]
RE the `Read timeout` issue -- we have some CI folks looking at this now. FYI

**Comment by tldahlgren** [comment]
> RE the `Read timeout` issue -- we have some CI folks looking at this now. FYI

FWIW.  I have also seen `petsc-3.20.0` timing out for the `e4s neoverse_v1` stack:

- https://github.com/spack/spack/pull/40827 (https://gitlab.spack.io/spack/spack/-/pipelines/538316/failures)

**Comment by v-dobrev** [comment]
Here's what I think is happening in the CI failure for `dealii`:
* CI tries to build `dealii` with `+arborx` and `+cuda`
* `dealii` does not propagate the `cuda` variant to `arborx`, so it is build without it: `^arborx~cuda`
* `arborx~cuda` depends on `kokkos~cuda`
* Spack adds the Kokkos paths to the environment when building `dealii`
* `dealii`'s CMake automatically finds Kokkos from the environment and tries to use it
* `dealii`'s CMake sees that the user wants to build with CUDA but Kokkos is build without CUDA, so it produces an error.

Some options to resolve this are:
* extend the CI spec of `dealii` to enable `arborx` with CUDA: `^arborx+cuda cuda_arch=80`
* update the `dealii` package to propagate the `+cuda` variant and `cuda_arch` to `arborx`
* suppress the auto-detection of Kokkos by `dealii` -- there's probably a suitable CMake variable that allow for that.

ping: @masterleinad

**Comment by luca-heltai** [comment]
There is an open PR for deal.II which is currently under investigation:  #40747. I think as soon as that is merged, things should work fine. Some things have changed in 9.5.0 and 9.5.1, and adding the checksum is not enough.

**Comment by balay** [comment]
https://gitlab.spack.io/spack/spack/-/jobs/8977525

```console
Executing: /opt/cray/pe/mpich/8.1.25/ofi/cray/10.0/bin/mpicc  -o /tmp/petsc-6lg_vrq_/config.setCompilers/conftest    /tmp/petsc-6lg_vrq_/config.setCompilers/conftest.o 
Testing executable /tmp/petsc-6lg_vrq_/config.setCompilers/conftest to see if it can be run
Executing: /tmp/petsc-6lg_vrq_/config.setCompilers/conftest
Running Executable with threads to time it out at 60
Executing: /tmp/petsc-6lg_vrq_/config.setCompilers/conftest
stdout: /tmp/petsc-6lg_vrq_/config.setCompilers/conftest: error while loading shared libraries: libefa.so.1: cannot open shared object file: No such file or directory
ERROR while running executable: Could not execute "['/tmp/petsc-6lg_vrq_/config.setCompilers/conftest']":
/tmp/petsc-6lg_vrq_/config.setCompilers/conftest: error while loading shared libraries: libefa.so.1: cannot open shared object file: No such file or directory
```
is this an issue on the CI test box side? presumably this build was functional before [when this version of petsc was added to spack]

**Comment by balay** [comment]
> There is an open PR for deal.II which is currently under investigation: #40747. I think as soon as that is merged, things should work fine. Some things have changed in 9.5.0 and 9.5.1, and adding the checksum is not enough.

yes - I'm waiting for that PR to get merged in. What I have here is a place holder until then [for dealii]. I guess I should replace it with the change from that PR - to better test dealii changes wrt xsdk.

**Comment by masterleinad** [comment]
@v-dobrev The latest failure for deal.II is described in https://github.com/spack/spack/pull/40747#issuecomment-1793521680. Looks like `nvcc`/`nvcc_wrapper` doesn't understand response file syntax. Do you have any idea how to resolve this?

**Comment by v-dobrev** [comment]
> @v-dobrev The latest failure for deal.II is described in [#40747 (comment)](https://github.com/spack/spack/pull/40747#issuecomment-1793521680). Looks like `nvcc`/`nvcc_wrapper` doesn't understand response file syntax. Do you have any idea how to resolve this?

@masterleinad, it looks like `nvcc` should work with response files, according to this: https://gitlab.kitware.com/cmake/cmake/-/merge_requests/3399. If there is a way to use directly `nvcc`, instead of `nvcc_wrapper`, when building deal.II that may resolve the issue.

**Comment by balay** [comment]
> If there is a way to use directly nvcc, instead of nvcc_wrapper

I do avoid nvcc_wrappers  in petsc build. But perhaps the issue is not the same.. [I need mpi wrappers to use Cxx compiler - not nvcc].

Here I use:

```python
    def revert_kokkos_nvcc_wrapper(self):
        # revert changes by kokkos-nvcc-wrapper                                                                                                                                                                                               
	if self.spec.satisfies("^kokkos+cuda+wrapper"):
            env["MPICH_CXX"] = env["CXX"]
            env["OMPI_CXX"] = env["CXX"]
            env["MPICXX_CXX"] = env["CXX"]

    def configure(self, spec, prefix):
        self.revert_kokkos_nvcc_wrapper()
        python("configure", "--prefix=%s" % prefix, *self.configure_options())
```

**Comment by eugeneswalker** [comment]
Sundials 6.6.2 is failing across the board on all platforms (arm, x86_64, power):
* [ sundials@6.6.2 /3pfva3p %gcc@11.4.0 arch=linux-ubuntu20.04-x86_64_v3 E4S ](https://gitlab.spack.io/spack/spack/-/jobs/8998640)
* ...
* [ sundials@6.6.2 /agqt6uc %gcc@11.4.0 arch=linux-ubuntu20.04-x86_64_v3 E4S ROCm External ](https://gitlab.spack.io/spack/spack/-/jobs/8997273)
```
==> Installing sundials-6.6.2-agqt6ucmoo24c463b42fbruey5ucob54 [1/1]
==> Fetching https://github.com/LLNL/sundials/releases/download/v6.6.2/sundials-6.6.2.tar.gz
==> No patches needed for sundials
==> sundials: Executing phase: 'cmake'
==> sundials: Executing phase: 'build'
==> sundials: Executing phase: 'install'
==> Error: ProcessError: Command exited with status 2:
...
 >> 7665    CMake Error at examples/arkode/C_serial/cmake_install.cmake:459 (f
             ile):
     7666      file INSTALL cannot find
     7667      "/tmp/root/spack-stage/spack-stage-sundials-6.6.2-agqt6ucmoo24c4
             63b42fbruey5ucob54/spack-src/examples/arkode/C_serial/ark_kepler_-
             -stepper_SPRK_--step-mode_fixed_--method_ARKODE_SPRK_EULER_1_1_--t
             f_50_--check-order_--nout_1.out":
     7668      No such file or directory.
     7669    Call Stack (most recent call first):
     7670      examples/arkode/cmake_install.cmake:47 (include)
     7671      examples/cmake_install.cmake:47 (include)
     7672      cmake_install.cmake:94 (include)
     7673    
     7674    
  >> 7675    make: *** [Makefile:113: install] Error 1
```

**Comment by masterleinad** [comment]
> I do avoid nvcc_wrappers in petsc build. But perhaps the issue is not the same.. [I need mpi wrappers to use Cxx compiler - not nvcc].

We need to do `nvcc_wrapper` since all files potentially contain `Kokkos` code. What I don't understand is where these response files come in and if they are used everywhere in spack or if some package forces them upon us or if `CMake` decides to use them. So far I have not been able to reproduce the failure on my end using `ninja` and `nvcc_wrapper` to configure deal.II+Kokkos+Cuda.

**Comment by balay** [comment]
BTW: did `deal.II+Kokkos+Cuda` feature ever work before? Perhaps that test should be disabled in CI?

**Comment by balay** [comment]
> Sundials 6.6.2 is failing across the board on all platforms (arm, x86_64, power):

yeah - its reported at https://github.com/spack/spack/pull/40920

**Comment by masterleinad** [comment]
> BTW: did `deal.II+Kokkos+Cuda` feature ever work before? Perhaps that test should be disabled in CI?

I don't think that was ever tested before using `spack`. We only made `Kokkos` a required dependency for `deal.II+Cuda` in the latest release. We have (non-spack) CI for this setup, though, see https://github.com/dealii/dealii/blob/578fcb0c15df4b88c42f71d676ac29a93ae007f3/.github/workflows/linux.yml#L231-L415.

**Comment by balay** [comment]
> We only made Kokkos a required dependency for deal.II+Cuda in the latest release

Is CI picking this change in spack - and running that test? If so - perhaps it should be de-coupled from the "new version" PR - into a separate PR -  enabled it only when it can be made to work in spack?

**Comment by masterleinad** [comment]
> Is CI picking this change in spack - and running that test? If so - perhaps it should be de-coupled from the "new version" PR - into a separate PR - enabled it only when it can be made to work in spack?

I'd be happy to just disable CUDA support for that deal.II release in `spack` at this point and try fixing it later.

**Comment by v-dobrev** [comment]
@masterleinad, part of the problem with the deal.II failures in CI is probably due the very long paths they use which in turn force CMake (and/or Ninja) to use response files which fail with `nvcc_wrapper`. In a normal build on Polaris, your PR worked fine for me with the spec `dealii+cuda~gmsh cuda_arch=80`. To speedup the build, I disabled `gmsh` -- adding that may also be a part of the reason that CMake (and/or Ninja) starts to use response files, however, the very long paths used in Spack CI may be the main reason.

**Comment by balay** [comment]
https://gitlab.spack.io/spack/spack/-/jobs/9002857
```
Executing: /opt/cray/pe/mpich/8.1.25/ofi/cray/10.0/bin/mpicc  -o /tmp/petsc-s7xkrbrq/config.setCompilers/conftest    /tmp/petsc-s7xkrbrq/config.setCompilers/conftest.o 
Testing executable /tmp/petsc-s7xkrbrq/config.setCompilers/conftest to see if it can be run
Executing: /tmp/petsc-s7xkrbrq/config.setCompilers/conftest
Running Executable with threads to time it out at 60
Executing: /tmp/petsc-s7xkrbrq/config.setCompilers/conftest
stdout: /tmp/petsc-s7xkrbrq/config.setCompilers/conftest: error while loading shared libraries: libefa.so.1: cannot open shared object file: No such file or directory
ERROR while running executable: Could not execute "['/tmp/petsc-s7xkrbrq/config.setCompilers/conftest']":
/tmp/petsc-s7xkrbrq/config
```

This error persists. @eugeneswalker any idea how to deal with this CI regression issue?

**Comment by masterleinad** [comment]
> @masterleinad, part of the problem with the deal.II failures in CI is probably due the very long paths they use which in turn force CMake (and/or Ninja) to use response files which fail with `nvcc_wrapper`. In a normal build on Polaris, your PR worked fine for me with the spec `dealii+cuda~gmsh cuda_arch=80`. To speedup the build, I disabled `gmsh` -- adding that may also be a part of the reason that CMake (and/or Ninja) starts to use response files, however, the very long paths used in Spack CI may be the main reason.

Yes, that's what I suspect, too, but I don't know if there is a reliable way to avoid using response files in `spack`. I'm also surprised that we don't see this problem with other packages like `Trilinos`.

**Comment by balay** [comment]
All, this PR is waiting on merge of:
- https://github.com/spack/spack/pull/40874
- https://github.com/spack/spack/pull/40747
- https://github.com/spack/spack/pull/41004

However - please review `[xsdk: add version 1.0.0]` commit - and let me know if you have any suggested changes.

[This branch would be rebased - and ready for merge with this one commit - after all the above PRs are merged in]

**Review by luszczek — DISMISSED** [review]

**Review by masterleinad — APPROVED** [review]

**Review by luszczek — APPROVED** [review]

**Pull request merged**
